drivers/display/ssd1322.c

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2024 Lukasz Hawrylko
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT solomon_ssd1322
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ssd1322, CONFIG_DISPLAY_LOG_LEVEL);
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/display.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/mipi_dbi.h>
+#include <zephyr/kernel.h>
+
+#include "ssd1322_regs.h"
+
+#define BITS_PER_PIXEL 4
+#define PIXELS_IN_BYTE (8 / BITS_PER_PIXEL)
+
+struct ssd1322_config {
+	const struct device *mipi_dev;
+	struct mipi_dbi_config dbi_config;
+	uint16_t height;
+	uint16_t width;
+	uint16_t column_offset;
+};
+
+static inline int ssd1322_write_command(const struct device *dev, uint8_t cmd,
+					const uint8_t *buf, size_t len)
+{
+	const struct ssd1322_config *config = dev->config;
+
+	return mipi_dbi_command_write(config->mipi_dev, &config->dbi_config, cmd, buf, len);
+}
+
+static inline int ssd1322_write_data(const struct device *dev, uint8_t *buf, size_t len)
+{
+	const struct ssd1322_config *config = dev->config;
+	struct display_buffer_descriptor mipi_desc;
+
+	mipi_desc.buf_size = len;
+	mipi_desc.width = len * PIXELS_IN_BYTE;
+	mipi_desc.height = 1;
+	mipi_desc.pitch = mipi_desc.width;
+	return mipi_dbi_write_display(config->mipi_dev, &config->dbi_config, buf,
+				      &mipi_desc, PIXEL_FORMAT_MONO01);
+}
+
+static int ssd1322_blanking_on(const struct device *dev)
+{
+	return ssd1322_write_command(dev, SSD1322_BLANKING_ON, NULL, 0);
+}
+
+static int ssd1322_blanking_off(const struct device *dev)
+{
+	return ssd1322_write_command(dev, SSD1322_BLANKING_OFF, NULL, 0);
+}
+
+static int ssd1322_write(const struct device *dev, const uint16_t x, const uint16_t y,
+			 const struct display_buffer_descriptor *desc, const void *buf)
+{
+	const struct ssd1322_config *config = dev->config;
+	size_t buf_len;
+	int ret;
+
+	if (desc->pitch < desc->width) {
+		LOG_ERR("Pitch is smaller than width");
+		return -1;
+	}
+
+	buf_len = MIN(desc->buf_size, desc->height * desc->width / 8);
+	if (buf == NULL || buf_len == 0U) {
+		LOG_ERR("Display buffer is not available");
+		return -1;
+	}
+
+	if (desc->pitch > desc->width) {
+		LOG_ERR("Unsupported mode");
+		return -1;
+	}
+
+	LOG_DBG("x %u, y %u, pitch %u, width %u, height %u, buf_len %u", x, y, desc->pitch,
+			desc->width, desc->height, buf_len);
+
+	uint8_t cmd_data[2];
+
+	cmd_data[0] = config->column_offset + (x >> 2);
+	cmd_data[1] = config->column_offset + (((x + desc->width) >> 2) - 1);
+	ret = ssd1322_write_command(dev, SSD1322_SET_COLUMN_ADDR, cmd_data, 2);
+	if (ret < 0) {
+		return ret;
+	}
+
+	cmd_data[0] = y;
+	cmd_data[1] = (y + desc->height - 1);
+	ret = ssd1322_write_command(dev, SSD1322_SET_ROW_ADDR, cmd_data, 2);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = ssd1322_write_command(dev, SSD1322_ENABLE_RAM_WRITE, NULL, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	uint8_t *buff_ptr = (uint8_t *)buf;
```

Comment:
buf is a buffer pointer, maybe data_ptr is better. You do not need explicit conversion here. Variable definition should be at the top of a block.